### DISCUSSION TITLE
Proper way of populating the nested objects default fields

### DISCUSSION THREAD

**Discussion by wojciechbator** [Kamva/mgm#39] [Q&A] [answered]
Hi! I use this library as an ODM in my project. I have an issue however with populating nested objects default model fields, like _id, created_at and updated_at. I persist the object using `Create()` and the root objects default fields are populated well, but the nested ones are missing. The nested docs are declared as structs that use ` mgm.DefaultModel `bson:",inline"` ` usual syntax. But, when persisted, the ids are missing and dates are like `"created_at" : ISODate("0001-01-01T00:00:00Z"), "updated_at" : ISODate("0001-01-01T00:00:00Z")`. I presume that it's because `Create` only supports 1st level, there's no nested population. What is the proper way of populating these fields in nested arrays? Do you provide ObjectID Factory in the library?

To visualize it a bit better:

```
struct Team {
  mgm.DefaultModel
  ...
  Players [{
    mgm.DefaultModel
    ...
  }]
```
and saving: `mgm.coll('teams').Create(TeamWithPlayers)`

results in `_id`, `created_at`, `updated_at` fields set properly only for team, not for players. Players are missing `_id` and the dates are as above.

**Comment by mehran-prs** [comment, marked as answer]
Hi @wojciechbator,
Yes, that's because `Create` only supports the first level, For nested levels, you need to use custom hooks.
For `_id` you can just simply use the `primitive.NewObjectID()` to generate a new ID for you.

This is what our default mgm model's hooks do:
```golang
// Creating hook is used here to set the `created_at` field
// value when inserting a new model into the database.
func (f *DateFields) Creating() error {
	f.CreatedAt = time.Now().UTC()
	return nil
}

// Saving hook is used here to set the `updated_at` field 
// value when creating or updateing a model.
func (f *DateFields) Saving() error {
	f.UpdatedAt = time.Now().UTC()
	return nil
}
```

You can change your root model's hooks to something like this:

```golang

type Player struct {
	mgm.DefaultModel
	// other fields...
}

type Team struct {
	mgm.DefaultModel
	Players []Player
	// other fields...
}

// Creating called when creating a new model.
func (t *Team) Creating() error {
	now := time.Now().UTC()

	t.CreatedAt = now
	for _, p := range t.Players {
		p.ID = primitive.NewObjectID()
		p.CreatedAt = now
	}
	
	return nil
}

// Saving when creating or updating a model.
func (t *Team) Saving() error {
	now:=time.Now().UTC()
	t.UpdatedAt = now
	for _, p := range t.Players {
		p.UpdatedAt = now
	}	
	
	return nil
}

```

**Reply by wojciechbator** [reply]
works fine, thanks :)

**Reply by mehran-prs** [reply]
You're welcome :)

**Comment by amirhossein-shakeri** [comment]
How about adding an embeded document model? something similar to `mgm.DefaultModel` for example `mgm.EmbeddedDocument` which is used in embeded documents/models & it has an embedded behavior! sounds cool! should I open an issue to request this feature or something? Its just an idea ...

**Comment by mehran-prs** [comment]
Hi @amirhossein-shakeri-work 
Embedded structs are the same as [this too](https://github.com/Kamva/mgm/discussions/39#discussioncomment-676050).
You need to define a webhook if you need to handle multiple things when a webhook is called.
